ipaclient/install/ipa_client_samba.py

```diff
@@ -0,0 +1,726 @@
+#
+# Copyright (C) 2019  FreeIPA Contributors see COPYING for license
+#
+# Configure the Samba suite to operate as domain member in IPA domain
+
+from __future__ import print_function
+
+import logging
+import sys
+import os
+import gssapi
+from urllib.parse import urlsplit
+from optparse import OptionParser  # pylint: disable=deprecated-module
+from contextlib import contextmanager
+
+from ipaclient import discovery
+from ipaclient.install.client import (
+    CLIENT_NOT_CONFIGURED,
+    CLIENT_ALREADY_CONFIGURED,
+)
+from ipalib import api, errors
+from ipalib.install import sysrestore
+from ipalib.util import check_client_configuration
+from ipalib.request import context
+from ipapython import ipautil
+from ipapython.errors import SetseboolError
+from ipapython.ipa_log_manager import standard_logging_setup
+from ipapython.dnsutil import DNSName
+from ipaplatform.tasks import tasks
+from ipaplatform.paths import paths
+from ipaplatform.constants import constants
+from ipaplatform import services
+from ipapython.admintool import ScriptError
+from samba import generate_random_password
+
+logger = logging.getLogger(os.path.basename(__file__))
+logger.setLevel(logging.DEBUG)
+
+
+@contextmanager
+def use_api_as_principal(principal, keytab):
+    with ipautil.private_ccache() as ccache_file:
+        try:
+            old_principal = getattr(context, "principal", None)
+            name = gssapi.Name(principal, gssapi.NameType.kerberos_principal)
+            store = {"ccache": ccache_file, "client_keytab": keytab}
+            gssapi.Credentials(name=name, usage="initiate", store=store)
+            # Finalize API when TGT obtained using host keytab exists
+            if not api.isdone("finalize"):
+                api.finalize()
+
+            # Now we have a TGT, connect to IPA
+            try:
+                if api.Backend.rpcclient.isconnected():
+                    api.Backend.rpcclient.disconnect()
+                api.Backend.rpcclient.connect()
+
+                yield
+            except gssapi.exceptions.GSSError as e:
+                raise Exception(
+                    "Unable to bind to IPA server. Error initializing "
+                    "principal %s in %s: %s" % (principal, keytab, str(e))
+                )
+        finally:
+            if api.Backend.rpcclient.isconnected():
+                api.Backend.rpcclient.disconnect()
+            setattr(context, "principal", old_principal)
+
+
+def parse_options():
+    usage = "%prog [options]\n"
+    parser = OptionParser(usage=usage)
+    parser.add_option("--server", dest="server", help="FQDN of IPA server")
+    parser.add_option(
+        "--netbios-name",
+        dest="netbiosname",
+        help="NetBIOS name of the machine",
+        default=None,
+    )
```

Comment:
It doesn't look to me that the server option determines which server the client connects to.